core/src/main/java/io/grpc/MessageEncoding.java

```diff
@@ -108,36 +118,93 @@ public InputStream decompress(InputStream is) throws IOException {
 
   /**
    * Registers a decompressor for both decompression and message encoding negotiation.
+   *
+   * @param d The decompressor to register
+   * @param advertised If true, the message encoding will be listed in the Accept-Encoding header.
    * @throws IllegalArgumentException if another compressor by the same name is already registered.
    */
-  public static final void registerDecompressor(Decompressor d) {
-    Decompressor previousDecompressor = decompressors.putIfAbsent(d.getMessageEncoding(), d);
-    checkArgument(previousDecompressor == null,
-        "A decompressor was already registered: %s", previousDecompressor);
+  @ExperimentalApi
+  public static final void registerDecompressor(Decompressor d, boolean advertised) {
+    DecompressorInfo previousInfo =
+        decompressors.putIfAbsent(d.getMessageEncoding(), new DecompressorInfo(d, advertised));
+    if (previousInfo != null) {
+      throw new IllegalArgumentException(
+          "A decompressor was already registered: " + previousInfo.decompressor);
+    }
   }
 
   /**
    * Provides a list of all message encodings that have decompressors available.
    */
-  public static Collection<String> getKnownMessageEncodings() {
+  @ExperimentalApi
+  public static Set<String> getKnownMessageEncodings() {
```

Comment:
Or you could just get rid of the other interfaces and just use `Codec`.  I know we had discussed this before ... maybe this is another argument in favor?